proposals/2024-01-29_native_histograms_text_format.md

```diff
@@ -0,0 +1,110 @@
+hist_with_classic_buckets_bucket{le="0.001"} 4
+hist_with_classic_buckets_bucket{le="+Inf"} 24
+hist_with_classic_buckets_count 24
+hist_with_classic_buckets_sum 100
+```
+The metric will have no "magic" suffixes, then the value for each series is a custom struct format with the following fields:
+* `sum: float64` - The sum of all observations for this histogram. Could be negative in cases with negative observations.
+* `count: uint64` - The number of samples that were observed for this histogram.
+* `schema: int32` - The schema used for this histogram, currently supported values are -4 -> 8.
+* `zero_threshold: float64` - The width of the zero bucket.
+* `zero_count: uint64` - The number of observations inside the zero bucket.
+* `negative_spans: []BucketSpan` - The buckets corresponding to negative observations, optional.
+* `negative_deltas: []int64` - The delta of counts compared to the previous bucket. 
+* `positive_spans: []BucketSpan` - The buckets corresponding to negative observations, optional.
+* `positive_deltas: []int64` - The delta of counts compared to the previous bucket. 
+
+A bucket span is the combination of an `int32` offset and a `uint32` length. It is encoded as `<offset>:<length>`. Lists/arrays are encoded within square brackets with elements separated by commas. Compared to JSON this avoids consistently repeating keys and curly braces.
+
+Positive infinity, negative infinity, and non number values will be represented as case insensitive versions of `+Inf`, `-Inf`, and `NaN` respectively in any field. This is the same behavior for values in OpenMetrics today.
+
+Note that in this initial implementation float histograms are not supported. Float histograms are rarely used in exposition, and OpenMetrics does not support classic float histograms either. Support could be added in the future by adding fields for `count_float`, `zero_count_float`, `negative_counts`, and `positive_counts`.
+
+### Backwards compatibility and semantic versioning
+
+After discussions with a few people it is believed that these changes can be made in a 1.x release of OpenMetrics. OpenMetrics 1.x parsers that support native histograms will still be able to read OpenMetrics 1.0 responses, therefore this change is backwards compatible. However, this change is not forwards compatible, i.e. an OpenMetrics 1.0 parser will not be able to read an OpenMetrics >= 1.1 response. Any producers implementing native histograms MUST also implement content negotiation and fall back to OpenMetrics 1.0.0, and therefore not expose native histograms, if a supported version cannot be negotiated. Note that the behavior to fall back to 1.0.0 is already part of the [OpenMetrics spec](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#protocol-negotiation).
```

Comment:
> After discussions with a few people it is believed that these changes can be made in a 1.x release of OpenMetrics. OpenMetrics 1.x parsers that support native histograms will still be able to read OpenMetrics 1.0 responses, therefore this change is backwards compatible. However, this change is not forwards compatible, i.e. an OpenMetrics 1.0 parser will not be able to read an OpenMetrics >= 1.1 response

Hm, I think this text misunderstands backward compatibility. To me, if 1.0 parser fails to read 1.x format (instead of e.g. skipping native histograms), that means native histograms in 1.x format is NOT backward compatible change. I would say let's be explicit about this. This change is NOT backward compatible, so the content negotiation mechanism HAS to be added to release it in 1.x (and that's fine). 

```suggestion
Implementing this change in 1.x would be considered a breaking change, because it could break 1.0 parsers. For this reasons this feature has to be behind a content negotiation. Any producers implementing native histograms MUST also implement content negotiation and fall back to OpenMetrics 1.0.0, and therefore not expose native histograms, if a supported version cannot be negotiated. Note that the behavior to fall back to 1.0.0 is already part of the [OpenMetrics spec](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#protocol-negotiation).
```

Reply:
I disagree, I think the behavior you describe is [forward compatibility](https://en.wikipedia.org/wiki/Forward_compatibility). Input designed for a 1.1 parser does not need to be handled by a 1.0 parser to classify as backwards compatible, the OpenMetrics spec makes sure of this by requiring all producers to fall back to 1.0.0 if they are not explicitly told otherwise.

Maybe this is semantics as to what a "breaking change" is though in the OpenMetrics world. If I see text such as "Implementing this change in 1.x would be considered a breaking change" I would expect this to require a major version bump which I don't think is what we want, and would also mean it is impossible to add anything to OpenMetrics 1.x as the format is incredibly locked down.

Reply:
I guess I have been thinking about it similar to Go compiler version changes. I don't expect code written for Go 1.22 to work with Go 1.0, but I do expect code written for Go 1.0 to still work with Go 1.22. I think of Go as backwards compatible so also think that this should be backwards compatible.

Reply:
Hehehe, @bwplotka and I nerded out about this on his podcast. I also find it hard to clearly define what is forward and backward compatibility. I also understood OM v1 formulated in a way that there isn't much you could change in a v1.1. However, the view expressed in this design doc makes way more sense to me. And reading the section about content negotiation in OM from this perspective suddenly also felt like it was meant that way by OM. However, discussions in the past lead to inconclusive results about this.

I would like most if we adopted the view expressed in the design doc, accepting that the current OM spec can be interpreted this way, but we will also clarify the intent in OM 1.1.

Reply:
Does it sound good to optimistically move forward with the view in this design doc, and if we receive pushback when creating the OM 1.1 changes we can find new consensus then if required?

Reply:
Shameless link to the mentioned episode with compatibility discussions: https://open.spotify.com/episode/2TPgtGv3boEegKQI9nTeE5?si=rMRhzSZ1T2aL_kjULPTZeA 🙈

Reply:
Thanks! After reading all here, I still think my suggestion makes sense. First and foremost, because I drop mention what's backward and what's forward, so we should all agree on this. (: Do you mind double-checking?

We can argue on what's backward, what's forward in separate place ;p

Reply:
In separation to my suggestion, in terms of what backward what's forward, I spent some time thinking about it and it feels it depends on the **point of view.**, or what is the **subject**.

### Details

Sounds like "the view expressed by this design doc is":

* forward compatibility = OM 1.0 parser works with OM 1.1
* backward compatibility = OM 1.x parser works with OM 1.0

If we replace C as client and P as protocol:

* forward: old C <-> new P 
* backward: new C <-> old P

This could make sense if we version parsers here (clients), so let's call it "Client View". But do we talk about parser compatibility or protocol in this section?

To me, in this context, it's opposite, and it's also how we formulated remote write spec compatibility e.g. [here](https://deploy-preview-2462--prometheus-docs.netlify.app/docs/specs/remote_write_spec_2_0/#backward-and-forward-compatibility). We put protocol in the center: Backward compatibility exists if new spec changes (e.g. OM 1.1) can be read by OM 1.0 client/parser without trouble (while skipping newest features). Forward compatibility in my mind is OM 1.1 client reading OM 1.0 which is easier to maintain, thus rarely mentioned in practice. Let's call this "Protocol View".

In our new convention, the Protocol View means:

* forward: old P <-> new C
* backward: new P <-> old C

This is perhaps also what SemVer assumes by saying [`MINOR version when you add functionality in a backward compatible manner`](https://semver.org/) in my mind. This means old C will work with new P (e.g. new dep). The new C working with old P (e.g. dep) (forward compatibility here?) does not make sense in this "upgrade" context.

To me, it seems like generally it's used both ways and it depends on the point of view e.g. do you see it from the point of view of the protocol or API ("new protocol" can be consumed by "old client"), or client ("new client" can consume "old protocol"). E.g. check [the wikipedia definition](https://en.wikipedia.org/wiki/Backward_compatibility)

> In [telecommunications](https://en.wikipedia.org/wiki/Telecommunications) and [computing](https://en.wikipedia.org/wiki/Computing), backward compatibility (or backwards compatibility) is a property of an [operating system](https://en.wikipedia.org/wiki/Operating_system), software, real-world product, or [technology](https://en.wikipedia.org/wiki/Technology) that allows for [interoperability](https://en.wikipedia.org/wiki/Interoperability) with an older [legacy system](https://en.wikipedia.org/wiki/Legacy_system), or with [input](https://en.wikipedia.org/wiki/Input/output) designed for such a system.

You can say both:

A) Backward compatibility is when "new client works for older API/protocol", so new C <-> old P
B) Backward compatibility is when "new API/protocol works for older client", so new P <-> old C

Both match definition kind of? This is similar in Go:

> I don't expect code written for Go 1.22 to work with Go 1.0, but I do expect code written for Go 1.0 to still work with Go 1.22. I think of Go as backwards compatible so also think that this should be backwards compatible.

First of all, funny enough the https://go.dev/doc/go1compat does not mention "backward" or "forward" prefix anywhere 🙈 But indeed everyone says "Go is backward compatible".

Secondly you assume `code written for Go 1.0` is `old C` and Go 1.22 SDK is `new P` and you expect that to work, so `old C <-> new P` is expected, and what's expected =  backward compatibility, which supports "protocol view", no?

In other words, to clone this to OM, it's backward compatibility when old parser (aka code written for Go 1.0) works with new protocol (Go 1.22 SDK)? This proves that one can assume client vs protocol as opposite things, and semantically it can still work.

## Summary

Both definition can technically work, but in my mind we change (and version!) protocol here, so P means protocol, not client here. It's a breaking change if old C does not work with new P with minor version bump (aka OM 1.0 parser with OM 1.1 spec).

From this POV, forward compatibility is when new C that supports new P cannot possibly work with old P, which is .. hard to even achieve in this context?

Interestingly enough we switch the definition if we talk about [senders "being backward" compatible with old receivers (via old protocol)](https://deploy-preview-2462--prometheus-docs.netlify.app/docs/specs/remote_write_spec_2_0/#backward-and-forward-compatibility:~:text=When%20talking%20to%201.x%20Receiver%2C%20Sender%20SHOULD%20use%20Content%2DType%3A%20application/x%2Dprotobuf%20for%20backward%20compatibility.), which proves both direction can be useful and can be mixed... as long as you are explicit/clear what is your main subject.

Reply:
Thanks for the interesting thoughts. It feels to me like actual experts on wire formats should have already found a clear definition for all of this. Do we have such an expert among us? It would be best to go with the view that already is a consensus among experts rather than coming up with something on our own (even though it might be self-consistent).

My preliminary conclusion is that OpenMetrics _might_ have in mind what @csmarchbanks explained in more detail in this PR. 
Of course, we could also define that an old OM parser must be able to parse any OM 1.x protocol, but given the terse way OM 1.0 is defined, I doubt we could ever have a 1.1. Which is another hint to me that OM always had @csmarchbanks's view in mind (but I agree that it wasn't clearly explained because I assumed initially that OM 1.1 is essentially impossible).

Reply:
@RichiH or @SuperQ perhaps one of you could weigh in on the intention of OpenMetrics versioning?

Reply:
My concern with your suggestion is that writing that this proposal is a breaking change makes me think it should be in 2.0 regardless of content negotiation (content negotiation doesn't allow us to break semver). The crux of the issue for me is whether we can support this feature in 1.1 or are forced to do a major version bump. If we think of this change as breaking then it should be put into 2.0 IMO. Following onto that, if we think this needs a major version bump we will need significantly more care in what else we include with a release, and think more about how to allow future changes without another major version bump (e.g. adding in a way to add new fields like what protobuf has, but does not exist in OM 1.0).

If we (and the OpenMetrics team) can agree that this can be released in 1.x, I think we could find some phrasing that works for both of us. I am not attached to the backwards/forwards compatibility wording, and agree that it depends on the point of view of the producers/consumers (in OpenMetrics parlance) vs of the protocol itself.

As @beorn7 mentions I think the only way we can release this (and any other new feature) in OM 1.x is if we take the client viewpoint. The client viewpoint is also what allows version content negotiation to be effective as content negotiation (for a version, clearly content negotiation is useful to negotiate text vs protobuf) is not helpful if all clients MUST be able to read all versions of a protocol anyway. This is the reason I think(hope?) the client view was possibly taken by the OpenMetrics team, but let's wait for some of them to respond :).

> Secondly you assume `code written for Go 1.0` is `old C` and Go 1.22 SDK is `new P` and you expect that to work, so `old C <-> new P` is expected, and what's expected = backward compatibility, which supports "protocol view", no?

Other way around is how I was thinking about it, the code written for Go 1.0 would be `old P` (the code is the text format protocol in this comparison), Go 1.22 compiler is `new C` as it is the new client consuming the protocol. Therefore `new C <-> old P`.